### DISCUSSION TITLE
Running Ulauncher v6 on demand (without a background daemon)

### DISCUSSION THREAD

**Discussion by friday** [Ulauncher/Ulauncher#1388] [Announcements]
After major refactoring and performance improvements, Ulauncher can be run without a background process (daemon).

To opt out of the background daemon, go to preferences:
- If you have systemd autostart available, disable "Launch at login".
- If you don't have systemd autostart, the autostart row shows a "Keep running in background" toggle instead. Disable that.

Ulauncher now won't use any system resources at all when it's not showing.

#### This comes with some caveats:
1. Having the background daemon running ready to show the window makes it possible to show it instantly, while opening the app from scratch will take longer. In my initial tests it started fast, but it was slow enough in sub-optimal conditions to sometimes not register my first keypress, which is a problem. And as much as we tried to optimize, we haven't gotten around this issue.
2. Extensions that rely on being able to perform scheduled/background tasks while idle (`ulauncher-timer` might be the only one that actually needs this) will not be able to do that in this mode.
3. If Ulauncher in the future adds supports for more features that requires performing background tasks, they will break when using this mode.
4. The "copy" action can fail, because it relies on a GTK method which is does some background action we can't get a callback for and know when it has completed, and if we close the app before it has completed it will fail. We have a workaround for this by forcing the app to be kept alive 1s after you copy. Let us know if this is not enough on your system.
5. `gapplication launch io.ulauncher.Ulauncher` may or may not work (see #1494), and if not, then `ulauncher-toggle` also will not work since it's just a shortcut to run `gapplication launch`. You can get around this by binding the keybinder to use `ulauncher` directly instead if it doesn't (I still haven't figured out why it works and not, but there is no benefit to using the gapplication way to launch Ulauncher with this mode).

Because of these caveats we **don't** recommend running without a daemon, but if you really care about your system idle use it may be worth it.

Note: This setting used to be called "Daemonless mode" and was an opt-in toggle under advanced settings. [#1721](https://github.com/Ulauncher/Ulauncher/pull/1721) (`v6.0.0-beta32`) renamed it to `keep_alive` (default on, so the daemon runs by default), and the user-facing surface moved into the autostart row in general settings. Existing `daemonless: true` settings migrate automatically to `keep_alive: false`.

**Comment by gornostal** [comment]
Continuing our discussion from #1493...
I measured startup time. It's 450ms (average of ~10 runs) on my machine.
I have an SSD of course. I don't know if it's good or not. Came with my Thinkpad T480s.

So yeah, the difference is noticeable.

My steps:
I made sure ulauncher was not running, added a new key binding with `/path/to/ulauncher/bin/ulauncher`, measured time between 5th line in `bin/ulancher` and the finish of `UlauncherWindow.__init__`.

Oh, and I have set `frequent apps` to `0`. I also have 4 extensions installed.

**Reply by gornostal** [reply]
I would take a slightly different approach with this.
Instead of the daemonless mode, I would keep the UL process in the BG, implement an extension process manager that would run extensions as they are needed. And maybe keep running frequently used ones for a while.
In the end we could have something like Android/iOS process manager but maybe less complex :)

**Reply by gornostal** [reply]
Why would anyone care about XXX MB of RAM anyway? :slightly_smiling_face: 
Let ulauncher run in the BG. 
Sure, we need to optimize a few things for those who install 10+ extensions.

**Reply by friday** [reply]
> Instead of the daemonless mode, I would keep the UL process in the BG, implement an extension process manager that would run extensions as they are needed. And maybe keep running frequently used ones for a while.

We have a separate ticket for this, and there is no reason not to do both imo.

> Why would anyone care about XXX MB of RAM anyway? 🙂

By having this attitude you self-select the user base to only include people who don't care about the wasted resources, and I just don't agree with this. I prefer to use lightweight apps only and try to replace any that are more resource intense than needed. I don't run any apps written in slow languages or using Electron, except for when there is no options (like Ulauncher being the only Python app I use). Most users who have the same priorities probably use Rofi or similar launchers instead. I think this is a big part of why Rofi is more popular. Linux users in particular care about system resources.

I have had no plans to make Daemonless mode the main/only mode unless all the caveats can be fully resolved. This might be never.

**Reply by friday** [reply]
> I measured startup time. It's 450ms (average of ~10 runs) on my machine.

That's a lot slower than for me. For me it's around ~80-82ms now, but it was similar to you before the latest optimization.

> I have an SSD of course. I don't know if it's good or not. Came with my Thinkpad T480s.

It's certainly better than not having SSD. Maybe mine is faster. My laptop is actually 3-4 years old now, but I got a pretty fast SSD back then.

> So yeah, the difference is noticeable.

You mean between the Daemonless mode and not?

> I made sure ulauncher was not running, added a new key binding with `/path/to/ulauncher/bin/ulauncher`, measured time between 5th line in `bin/ulancher` and the finish of `UlauncherWindow.__init__`.

Actually, you should measure between the start and the `on_initial_draw`, when `__init__` is finished the window hasn't been drawn yet (at least for me). That should make the measurements slower though.

You can test adding a `time.sleep(1)` to various event listeners and see if that slows the app down from starting. If not, then it means the window is drawn before this, so it's safe to run the deferred logic.

> Oh, and I have set `frequent apps` to `0`. I also have 4 extensions installed.

How about "Start each session with a blank query"? Try enabling it.

Also, if you run `make run` you get the log output before Ulauncher starts. If it says anything about loading extensions before you type something, it means the deferring isn't working for you.

**Reply by gornostal** [reply]
Thank you! I'll try your suggestions.

> ...who don't care about the wasted resources

RAM is the only resource that is wasted if not used fully. 😉

OSes have many features that use RAM for better performance, such as file caching, memory-mapped files, and more. 
Unfortunately, Python with GTK requires loading a looot of files. `gir1.2` is probably the largest dependency. 

That's why I think keeping UL running is an improvement.

**Reply by friday** [reply]
> OSes have many features that use RAM for better performance, such as file caching, memory-mapped files

This is a reason for **applications** to use less ram, so that the **system** can manage itself better and file system caching will work as intended. But also that there will be lower risk of trashing, or the oom killer randomly killing your ram, or Slack breaking because of the trashing/swapping for example (as it frequently does).

**Reply by friday** [reply]
Once we fix https://github.com/Ulauncher/Ulauncher/issues/1063 Ulauncher will use a smaller and more consistent part of your ram of course. I think that's the bigger issue. I already had the same argument about this with "goodwillcoding" in 2018 and I wasn't buying it then either: https://github.com/Ulauncher/Ulauncher/issues/285 (if anything, this turned me off from contributing).
